Remove `email` trait requirement & support anonymous users

ryana opened the pull request:
Hi there,

We recently made two changes to Ramen that require changes to our integration library. First, we no longer require users to have an `email` attribute, we need to remove the check for that.

We also now support calling `identify()` with no traits (aka [Anonymous Users](https://ramen.is/blog/release-notes-anonymous-users-conditional-follow-ups-new-ramenjs-api-more/)), but we wanted to reinforce that our API would not be called until the user was `identify`'d, hence the addition of the `this.identified` check.

ryana committed:
Remove  trait requirement & support anonymous users

put component & History back. fix names of tests

checkpoint

Refactor integration to use new _ramen API & support anonymous users

Auto-identify on track/page/group

Address comment from Segment'ers last round of feedback

Add documentation comments

Remove the track event properties since Ramen doesn't process them

Make test names match the desired functionality

Changes per Segment's comments

hankim813 commented on test/index.test.js:
```diff
@@ -79,10 +79,9 @@ describe('Ramen', function() {
+        analytics.identify('12345', { email: 'ryan@ramen.is' });
         analytics.page();
-        analytics.called(window.Ramen.go);
+        analytics.calledTwice(window.Ramen.go);
```
Why are we changing this here? This would not make sense with the description of the title :/

hankim813 commented on test/index.test.js:
```diff
@@ -128,11 +126,8 @@ describe('Ramen', function() {
-
-        analytics.identify('id');
-        analytics.didNotCall(window.Ramen.go);
+        analytics.identify('this-users-id');
```
this test description should be updated then no?

hankim813 commented:
Overall I think this looks good -- few things though:
- I think we should add `analytics.called` in addition to what we have already (`.calledOnce`, `.calledTwice`) so that we can check that the proper argument was passed when `window.Ramen.go` is called. Right now, these tests are just checking that the function was invoked but not really making sure the right settings/payload were called, which makes these tests not all too helpful to prevent bugs down the road, especially since the `privateGo` function as a decent amount of logic in how it sets different variables.
- could you remove the changes to `History.md` and the `component.json` since that should be done when we make a release so we want to keep that separate from this PR.
- could you squash your commits?

Thanks for the PR btw!
pinging @f2prateek for another set of 👀 here

f2prateek commented:
Code LGTM.

Behaviour wise — Instead of simply dropping calls for anonymous users, shouldn't they be queued until the user is identified?

ryana commented:
Ok changes to component.json & History.md are out and commits squashed.

On the `called` part, I understand what you're saying: for the tests that I changed, I'm not testing any of the underlying behavior of the `privateGo` method. In other tests, however, I am verifying that `window.ramenSettings` is being set correctly.

Would you prefer me to rewrite all the tests to use `called` the same way it's used in [Intercom](https://github.com/segment-integrations/analytics.js-integration-intercom/blob/master/test/index.test.js) for example? Or would it suffice to continuing using the method I'm currently using where I verify that `ramenSettings` is being set and just add that to the tests that are currently only calling `called(Once|Twice)`?

Thanks so much to you both for the quick reply and the thoughtful commentary!

hankim813 commented:
@ryana ahh I'm sorry I missed those other tests! 

Haha nice
![](https://cloudup.com/cf3cs7PeluM+)

> Would you prefer me to rewrite all the tests to use called the same way it's used in Intercom for example? Or would it suffice to continuing using the method I'm currently using where I verify that ramenSettings is being set and just add that to the tests that are currently only calling called(Once|Twice)?

No you don't need to rewrite. You can test either way so no need to change it imo. Still achieves the same effect. Since you have separate tests actually making sure that the `ramenSettings` are set properly, no need to add that to the `calledOnce/Twice` again :)

I do want to hear your opinion on what @f2prateek said:

> Behaviour wise — Instead of simply dropping calls for anonymous users, shouldn't they be queued until the user is identified?

Because we could document that we will drop any events before the first `.identify()` call if Ramen simply does not support anonymous events. But with that said, just curious -- why not support anonymous events?

ryana commented:
Ok great thanks.

On the queue'd part, I just found this is in your documentation:

```
Note: You won’t need to call identify for anonymous visitors to your site. We’ll automatically assign them an anonymousId, so just calling page and track will still work just fine without identify.
```

I didn't realize that. So I guess in order to properly support anonymous users, we shouldn't be requiring `identify` to be called either -- for some reason I thought you required `identify()` to be called even for anonymous users.

Let me swing back to the whiteboard for a bit, but I think that we're on the same page, and that instead of queuing up the calls, we'll just let them propagate through and insert an automatic `identify()` on our end.

Thanks again for the quick reply!

hankim813 commented:
@ryana ahh yeah so `analytics.js` will automatically generate an `anonymousId` and save it to the Segment cookie. Every API call will have this `anonymousId` attached. 

If you call `.identify()` with a `userId`, all subsequent calls will have BOTH that same `anonymousId` and the passed in `userId`. So what many partners will do is either a) use their own anonymousId management and alias that to the `userId` when `.identify()` is called b) use Segment's anonymousId and userId to attribute anonymous events to the user.

So we always recommend customers to only `.identify()` when you can actually identify the user like after signup/login. Hope that helps! Let me know if you have any other questions about library behavior or anything like that :)

ryana commented:
Hey @hankim813 I'm working on refactoring these tests and I'm up against a crazy issue. You ever seen this?

```
  1) Ramen after loading #identify should call _ramen.identify with the id:
     Error: Expected "stub" to be called with "[
  {
    "id": "New"
  }
]", 
but it was called with "[
  {
    "id": "New"
  }
]".
```

Here's the code:

```
113     describe('#identify', function() {
114       beforeEach(function() {
115         analytics.stub(window._ramen, 'identify');
116       });
117 
118       it('should call _ramen.identify with the id', function() {
119         analytics.identify('New');
120         analytics.called(window._ramen.identify, {
121           id: 'New'
122         });
123       });
```

I'm befuddled.

ryana commented:
Figured this out. I was calling it with an Object that had `null` values for a few keys, and those keys were not being printed out in the error message.

hankim813 commented:
Haha I've run into those before... when in doubt I just `console.log` everywhere.

ryana commented:
Ok @hankim813 I've majorly refactored stuff :)

RamenJS's new release from a few months ago moved away from using `ramenSettings` and `Ramen.go()` and towards a safer `window._ramen` API that does the "I'm an Array until the library loads, and then I become an instance of window.Ramen.Boot" thing.

At first, I wanted to make this update small, so I kept it using that legacy API, but after getting in here I figured I'd just go all out :)

I also set it up so that `group`, `page`, and `track` will fire events before `identify` is called per @f2prateek's advice.

I tried to write my own `Ramen.prototype.reset()` so that `this.identified` would be set back to `false` but that made all the tests fail so I'm guessing maybe I'm not supposed to do that. Let me know.

Thanks!

hankim813 commented on lib/index.js:
```diff
@@ -46,187 +43,103 @@ Ramen.prototype.initialize = function() {
-Ramen.prototype.identify = function(identify) {
-  privateGo({ integration: this, identify: identify });
-};
+  var props = group.properties();
```
Can you change this to `group.traits()` since we use `traits` for group calls. Although functionally speaking this is okay since `.properties()` [looks up traits](https://github.com/segmentio/facade/blob/master/lib/group.js#L124) underneath anyway but just for the sake of keeping consistency across integrations :)

ryana replied:
Sure will do.

hankim813 commented on lib/index.js:
```diff
@@ -46,187 +43,103 @@ Ramen.prototype.initialize = function() {
-};
+  var props = group.properties();
+  props = alias(props, { createdAt: 'created' });
+  props = alias(props, { created: 'created_at' });
```
You can do this all in one line:

``` js
var props = group.traits({ createdAt: 'created', created: 'created_at' });
```

ryana replied:
Cool thanks. Will do.

hankim813 commented on lib/index.js:
```diff
@@ -46,187 +43,103 @@ Ramen.prototype.initialize = function() {
+  props = alias(props, { createdAt: 'created' });
+  props = alias(props, { created: 'created_at' });
+  var id = group.groupId();
+  if (id) props.id = id;
```
Nitpick here but why not `if (group.groupId()) props.id = group.groupId();` ;p if we are not going to use the variable `id` later on in this function

ryana replied:
No problem. I think I copied this pattern from another integration :)

hankim813 commented on lib/index.js:
```diff
@@ -46,187 +43,103 @@ Ramen.prototype.initialize = function() {
+    this.identify({ anonymous: true });
+  }
+
+  window._ramen.page();
```
question -- sometimes our users like to manually override the properties such as `url`, `title`, `path`, etc by pushing them inside the `.page()` call like

```
analytics.page({
   title: 'some custom title',
   path: 'some custom path'
});
```

Does Ramen allow you to override its own `.page()` call's auto parsing of the document by passing in arguments into the `window._ramen.page();`? If so we might want to allow support for that. But happy to make that a separate PR since this one already is a big overhaul :)

ryana replied:
Nope we just look at current URL when `_ramen.page()` is called.

hankim813 commented on lib/index.js:
```diff
@@ -46,187 +43,103 @@ Ramen.prototype.initialize = function() {
 
-  email = email || integration.analytics.user().traits().email;
-  id = id || integration.analytics.user().id();
+  this.identified = true;
```
should this be at the bottom of the function?

ryana replied:
Yup. Moving it.

hankim813 commented on lib/index.js:
```diff
@@ -46,187 +43,103 @@ Ramen.prototype.initialize = function() {
 
-  // We need email addresses
-  if (!email) {
+  if (identify.anonymous) {
```
Is this in relation to catching what you've been doing in the previous calls where you are calling `this.identify({ anonymous: true });`? Then you'd want this conditional to be `identify.traits().anonymous` -- since `this.identify()` will call `Ramen.prototype.identify`

ryana replied:
Removing all these checks in favor of @f2prateek's comment below.

hankim813 commented on lib/index.js:
```diff
@@ -46,187 +43,103 @@ Ramen.prototype.initialize = function() {
-    del(traits, 'created');
-    del(traits, 'createdAt');
-  }
+  user = { id: identify.userId() };
```
Note that if someone calls `analytics.identify({ someTrait: 'hello' });` aka without any `userId` which can definitely happen, this would be setting `user = { id: undefined }`

ryana replied:
Yikes. Ok what's the right way to get "application-provided ID or segment-provided ID"?

hankim813 commented on lib/index.js:
```diff
@@ -46,187 +43,103 @@ Ramen.prototype.initialize = function() {
 
-  // Expose ramenSettings so Ramen.go() can see it
-  window.ramenSettings = extend(ramenSettings, window.ramenSettings || {});
+  user.traits = alias(user.traits, { createdAt: 'created' });
```
Can use the similar aliasing as noted above:

```
user.traits = identify.traits({ createdAt: 'created', created: 'created_at' });
```

ryana replied:
Yup.

f2prateek commented:
I think the code will be simpler if you don't try to reuse the identify method from the group call.

```
Ramen.prototype.group = function(group) {
  if (!this.identified) {
    window._ramen.identify();
    this.identified = true;
  }

  var props = group.properties();
  props = alias(props, { createdAt: 'created' });
  props = alias(props, { created: 'created_at' });
  var id = group.groupId();
  if (id) props.id = id;

  window._ramen.group(props);
};
```

And you can remove the anonymous check in the identify method.

hankim813 commented:
Agreed with @f2prateek, also what does `window._ramen.identify()` effectively doing in your library? Is it safe to keep calling those in the same browser session?

ryana commented:
It is safe to call `window._ramen.identify()` multiple times. It's a no-op if nothing is changing.

hankim813 commented:
Awesome -- once you make the changes based on our inline comments happy to take a look again and ship this awesome update! :D

ryana commented on lib/index.js:
```diff
@@ -46,187 +43,101 @@ Ramen.prototype.initialize = function() {
-
-    del(traits, 'company');
+  if (!user.id) {
+    return;
```
@hankim813 I looked at some other libraries, and they will drop calls on the floor if no `id` is passed. I think I'll do that too, as `identify`ing a user with no `id` could cause some weird behavior inside Ramen. I'd rather stick to anonymous or known w/ `id`.

That cool?

hankim813 replied:
Yep this is completely fine if that will have adverse effects to using Ramen. 

Normally if people are calling `.identify()` without a userId, they are really just caching the `traits`. But anyway doing this is not what we recommend so it's an anti-pattern.  

Just make sure to document this requirement :)

ryana replied:
Ok cool then we're good to go!

I'll get working on updating our [docs page](https://gist.github.com/ryana/4bfe54d32ebcc229bba7) right now. Do I still just send that into partners@segment.com to get it updated?

ryana commented:
Ok I have one more question before we pull the trigger on this: should I setup `page()` and `track()` to send in attributes to `_ramen` even though Ramen ignores those data right now?

I can explicitly document [here](https://segment.com/docs/integrations/ramen/) that we don't accept those parameters, but in the event we start supporting that stuff, it would be great to not have to bother you again :)

The only downside to doing this is that it could confuse people who are looking at this source into thinking we do support those attributes.

Let me know what you think, and then I'll respond and we should be done.

Thanks!

hankim813 commented:
> Ok I have one more question before we pull the trigger on this: should I setup page() and track() to send in attributes to _ramen even though Ramen ignores those data right now?
> 
> The only downside to doing this is that it could confuse people who are looking at this source into thinking we do support those attributes.

Agreed. If you don't support them then I would not send the manual properties if it's not supported yet.

I'd just add a note in the docs that Ramen Js library auto parses attributes about the page :)

Last thing, could we add these:

``` js
/**
 * Identify.
 *
 * @api public
 * @param {Identify} identify
 */
```

For all the other functions like `.page()`, `.track()`, `.group()`? Just want to be consistent 👍 

Thanks!

ryana commented:
Okie dokie. Added comments. Removed passing attributes to `track` that were being ignored by Ramen. Updated tests. Squashed commits. Great success.

Here is our updated documentation: https://gist.github.com/ryana/4bfe54d32ebcc229bba7 -- Do you take that or should I email it to partners@segment.com?

hankim813 commented:
yep you can send it to partners@segment.com :)

@f2prateek let's run through the final code one last time before merging?

Awesome work on this @ryana !!

hankim813 commented:
@ryana one last run through and found these small things we can quickly change to make the code even tighter :) Once these changes are made I can merge and deploy!

Line 62-63 can basically be merged to:

``` js
var props = group.traits({ createdAt: 'created', created: 'created_at' });
```

The `.traits()` and `.properties()` function are baked in with aliasing feature if you pass in an object of alias fyi ;)

Let's move 120-122 to the top of the `.identify()` function since nothing will matter if this conditional fails.

``` js
Ramen.prototype.identify = function(identify) {
 // don't identify without userId since it would have adverse effects in Ramen (or if you have a better comment to add here on why we return out, would be helpful for our success engineers!)
  if (!identify.userId()) return;
  var user;
  var traits;
  var opts;
  ...
```

Otherwise LGTM!

ryana commented:
Done!

hankim813 commented:
@ryana just combing through the docs as I was talking to our partners team to ship the changes, could we change the following:

```
Note: By default, the RamenJS widget monitors the URL for changes, but that functionality is disabled by the Segment integration, necessitating calling page on your own.
```

That is under the `.page()` section. The `.page()` call is included in the Segment snippet so unless the user removed that (which most don't), you can assume that page is called on every page load so I don't think we need this note since it sounds like a restriction of Ramen's feature set by Segment which is not true :)

Also I would add under the `Identify` section that the call will not be process if you do not pass a `userId`. This should be something like the below so that it's nice and prominent:

**NOTE**: `.identify()` call without `userId` will be ignored since it will blah blah

ryana commented:
Cool thanks. Docs updated: https://gist.github.com/ryana/4bfe54d32ebcc229bba7

hankim813 commented:
@ryana awesome!! I'll pass this to our partners team to get the docs updated and get this shipped! will ping you again once all of that is done :) 

thanks again for the quick turnaround here!

hankim813 commented:
@ryana just shipped to production!
